### ISSUE TITLE
Subclass model can't restore weights from h5 format

### ISSUE THREAD

**Issue by vermouth1992** [tensorflow/tensorflow#42381]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.3.0-rc2-23-gb36436b087
- Python version: 3.7.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.2
- GPU model and memory: GTX 1060 6GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
The last two prints should print a tensor of random numbers.

**Describe the expected behavior**
The last two prints should print a tensor of ones. It works if we save it as TF format.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```python
import tensorflow as tf


class Model(tf.keras.Model):
    def __init__(self):
        super(Model, self).__init__()
        self.a = tf.Variable(initial_value=tf.random.normal([5]), trainable=False, name='a')
        self.call = tf.function(func=self.call, input_signature=[
            tf.TensorSpec(shape=[None, 5], dtype=tf.float32)
        ])
        self.prod = tf.function(func=self.prod, input_signature=[
            tf.TensorSpec(shape=[None, 5], dtype=tf.float32)
        ])
        self.set = tf.function(func=self.set, input_signature=[
            tf.TensorSpec(shape=[5], dtype=tf.float32)
        ])
        self.build(input_shape=tf.TensorShape(dims=[None, 5]))

    def call(self, inputs):
        print(f'Tracing call with inputs={inputs}')
        return self.a + inputs

    def prod(self, inputs):
        print(f'Tracing prod with inputs={inputs}')
        return self.a * inputs

    def set(self, value):
        print(f'Tracing set with inputs={value}')
        self.a.assign(value)


if __name__ == '__main__':
    model = Model()
    print(model(tf.zeros(shape=[2, 5])))
    print(model.prod(tf.ones(shape=[2, 5])))
    model.set(tf.ones(shape=[5]))
    print(model(tf.zeros(shape=[2, 5])))
    print(model.prod(tf.ones(shape=[2, 5])))
    print(model.weights)
    model.save_weights('/tmp/a.h5')
    del model
    model = Model()
    model.load_weights('/tmp/a.h5')
    print(model(tf.zeros(shape=[2, 5])))
    print(model.prod(tf.ones(shape=[2, 5])))

```
The bahavior persists in Google colab: https://colab.research.google.com/drive/1b8WuO_hBvYynY5gkH_MSI0FJS0Qrj3MF?usp=sharing

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@vermouth1992 
Please refer to [the gist](https://colab.research.google.com/gist/Saduf2019/2e14f6792a3cfd38b212ce114d7daaf2/untitled373.ipynb) and confirm if it replicated the issue reported.Thanks!

**Comment by vermouth1992** [comment]
Yes. The issue is replicated.

**Comment by vermouth1992** [comment]
> @vermouth1992
> Please refer to [the gist](https://colab.research.google.com/gist/Saduf2019/2e14f6792a3cfd38b212ce114d7daaf2/untitled373.ipynb) and confirm if it replicated the issue reported.Thanks!

Yes. The issue is replicated.

**Comment by jvishnuvardhan** [comment]
@vermouth1992 This is working as intended. When you have a custom_object and want to save model in 'h5' format, then you need to add a `get_config` method as shown below. This `get_config` is not required if you save the model in 'tf' format. Check [this guide](https://www.tensorflow.org/guide/keras/save_and_serialize#custom_objects) on TF website for more details.

```
    def get_config(self):
        config = super(Model, self).get_config()
        config.update({"a": self.a})
        return config
```


Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/629d3f22e24c7497c1473d1296eb71eb/tf_save_h5_bug.ipynb). Thanks!

Please verify once and close the issue if this was resolved for you. Thanks!

**Comment by vermouth1992** [comment]
> @vermouth1992 This is working as intended. When you have a custom_object and want to save model in 'h5' format, then you need to add a `get_config` method as shown below. This `get_config` is not required if you save the model in 'tf' format. Check [this guide](https://www.tensorflow.org/guide/keras/save_and_serialize#custom_objects) on TF website for more details.
> 
> ```
>     def get_config(self):
>         config = super(Model, self).get_config()
>         config.update({"a": self.a})
>         return config
> ```
> 
> Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/629d3f22e24c7497c1473d1296eb71eb/tf_save_h5_bug.ipynb). Thanks!
> 
> Please verify once and close the issue if this was resolved for you. Thanks!

I see the problem. I actually tried what you said but it didn't work because I declared the tf.Variable inside the class. I changed the declaration inside the class and it doesn't work. So is it correct that all the variables have to be passed in as an argument for h5 format to work?

**Comment by vermouth1992** [comment]
Also, is there any better solution if I would like to decorate a class member function with batch shape (e.g. [None, dim]), where dim is passed in as an argument from the class constructor. Currently, I simply decorate them inside the constructor but this is actually really ugly. This is very useful to avoid retracing if the batch size is always changing such as in reinforcement learning, where the length of the trajectory is always changing. Thank you very much for your help!

**Comment by jvishnuvardhan** [comment]
@vermouth1992 Defining `a` inside `def __init__(self):` is problematic as and when you instantiate the subclass model, it will `run def __init__(self):` and reinitialize ` a` with a `tf.random.normal([5])`. Thanks!

**Comment by jvishnuvardhan** [comment]
Please close the issue if this was resolved for you. Thanks!
